### PR TITLE
feat: calc defaultEnd based on unsorted array

### PR DESCRIPTION
**Pull request by xmxanuel**
This PR contains a new approach for calculating the `defaultEnd` based on an unsorted array.

All existing tests are passing and I added a couple more.

However, there might be still some edge cases the new approach doesn't cover.

We move the discussion into this PR.

### Problem Description
A drip can be optional streamed until a user runs out of funds. These drips are called `defaultDrips`.
They are defined by a start time and an `amtPerSec`.  The `duration=0` indicates the defaultEnd.

After our implementation covers all normally scheduled drips (with a start and duration) we end up with a  leftover `balance` and an array of these defaultDrips.

The leftover balance can be used for defaultDrips.

We need to calculate based on the start dates and amtPerSec the block.timestamp when the user leftover balance will equal zero. 

This timestamp is the point in the time, when all defaultDrips will run out of funds.

**Example**
```
Balance=100
DripA: start: 100, amtPerSec=2
DripB: start: 120, amtPerSec=4

Here the defaultEnd will be 130. From timestamp 100 to 120 we drip 20*2 = 40.
Afterward the second one starts and we have a total amtPerSec 6. 120 to 130 we drip 10*6 = 60;

At timestamp 130 the entire balance has been spent.

100                 120         130
|--------------------|----------|
      amt=2           amt=2+4
```

### Current Algorithm
The current approach is to first sort the default drips array. The algorithm then looks at the time difference between two drip start dates step by step and calculates the amount spend during this time period. The totalAmtPerSec is increased afterwards with a new amtPerSec.

In each step, a temporary end time is calculated based on the currently seen default drips if they can be covered until the next drip.

In the example from above, the duration between the first two drips is 20 with a amtPerSec=2.

However, this approach requires as to first sorting the array which is expensive. 

### New Design
It is not required to sort the array in the first place. The entire problem can be seen as an equation. Which describes when the spend drips amount will be equal to the balance.

```
x... defaultEnd
s... start
a... amount
b... balance
```
**Entire Balance Spend Equation**
Example for only two default drips

$$
a_1(x-s_1) + a_2(x-s_2) = b
$$

The duration of a default drip can be described as `defaultEnd-start` multiplied with the `amtPerSec` of the drip results in the total amount spend for one drip.  The sum needs to equal the balance `b`.

This equation can be solved for x (defaultEnd).

$$
x = \frac{b +  a_1*s_1 +  a_2*s_2}{ a_1  + a_2}
$$

$$
x = \frac{b +  a_1 s_1 +  a_2 s_2}{ a_1  + a_2}  
$$

**Default End Formular for n default drips**

$$
x = \frac{b +  a_1 s_1 + ... + a_n s_n}{ a_1 + ... + a_n}  
$$

### Trade-offs
This new design reduces the complexity of the code and will be more gas efficient. 

However, it requires a slightly change in the behavior of the system. 

In the current design the defaultDrips array can contain drips which can't be covered. This means the `defaultEnd` is smaller than the start time of the drips.

Currently, we would not add this uncovered default drips to the delta timeline. They are basically ignored when iterating over the list. 

In the new design, we would need to `revert` if a user adds new default start drips which can't be covered. 

### Open Questions
If our formula results in a defaultEnd timestamp that is smaller than the highest default start date. We are in a scenario, where not all default drips can be covered. In this case, we would revert.

- Is this assumption always true?
- Can we prove it?

### Implementation
- implements described algorithm from above
- removes the DefaultDrips struct and helper functions for generating the array

### CODE REVIEW

**Review by CodeSandwich — CHANGES_REQUESTED** [review]
I love it, so simple :+1:

File: src/Drips.sol [hunk 1]
```diff
@@ -292,13 +292,16 @@ library Drips {
     /// Must be sorted, deduplicated and without 0 amtPerSecs.
     /// @return defaultEndTime The end time of drips without duration.
     function _defaultEnd(uint128 balance, DripsReceiver[] memory receivers)
-        private
+        internal
```
[the remark this answers is not in the document]
Comment (xmxanuel) [reply src/Drips.sol:295]: I think it can be public. I changed it to test the function directly.

File: src/Drips.sol [hunk 2]
```diff
@@ -307,78 +310,28 @@ library Drips {
             // Default drips end doesn't matter here, the end time is ignored when
             // the duration is zero and if it's non-zero the default end is not used anyway
             (uint32 start, uint32 end) = _dripsRangeInFuture(receiver, _currTimestamp(), 0);
+            if (start > highest) highest = start;
             if (receiver.duration == 0) {
-                length = _addDefaultEnd(defaults, length, start, receiver.amtPerSec);
+                defaultSum += uint256(uint256(receiver.amtPerSec) * start);
+                defaultDivisor += uint256(receiver.amtPerSec);
```
[the remark this answers is not in the document]
Comment (xmxanuel) [reply src/Drips.sol:316]: :+1:

File: src/Drips.sol [hunk 3]
```diff
@@ -307,78 +310,28 @@ library Drips {
...
         }
-        if (end > type(uint32).max) end = type(uint32).max;
-        return uint32(end);
+        // main formular to calculate default end time
+        // a ...  amtPerSec
+        // s ...  start
+        // defaultEnd =(balance + a_1*s_1 + ... + a_n*s_n)/(a_1 + ... a_n)
+        uint136 defaultEnd = uint136((balance + defaultSum) / defaultDivisor);
+        if (defaultEnd > type(uint32).max) defaultEnd = type(uint32).max;
+        require(defaultEnd >= highest, "Run out of funds before default drips start");
```
[the remark this answers is not in the document]
Comment (xmxanuel) [reply src/Drips.sol:333]: I had it first like that.  Some of the existing unit tests failed. 
In general I agree with you. I will investigate it.

File: src/Drips.sol [hunk 4]
```diff
@@ -307,78 +310,28 @@ library Drips {
...
+        if (defaultDivisor == 0) {
+            return 0;
         }
-        if (end > type(uint32).max) end = type(uint32).max;
-        return uint32(end);
+        // main formular to calculate default end time
+        // a ...  amtPerSec
+        // s ...  start
+        // defaultEnd =(balance + a_1*s_1 + ... + a_n*s_n)/(a_1 + ... a_n)
+        uint136 defaultEnd = uint136((balance + defaultSum) / defaultDivisor);
```
[the remark this answers is not in the document]
Comment (xmxanuel) [reply src/Drips.sol:331]: Yes, it is `defaultEnd`.

The duration can be defined as: `defaultEnd-start = duration`. That's the main assumption of the formular

File: src/Drips.sol [hunk 5]
```diff
@@ -292,13 +292,16 @@ library Drips {
...
         returns (uint32 defaultEndTime)
     {
         require(receivers.length <= MAX_DRIPS_RECEIVERS, "Too many drips receivers");
-        DefaultEnd[] memory defaults = new DefaultEnd[](receivers.length);
-        uint8 length = 0;
+        // helper variables for the the default formular
+        uint256 defaultSum = 0;
+        uint256 defaultDivisor = 0;
+        // highest seen start date
+        uint32 highest = 0;
```
[the remark this answers is not in the document]
Comment (xmxanuel) [reply src/Drips.sol:304]: It is hard to name those things, they are just part random parts of the formula grouped together to optimize the calculation. 

One part is the helper for the divisor the other one for the sum. 

I think it is a personal taste for `defaultsHighestStart`. If the definition is just one line above with a good comment you can also call it just `h`.  Then the brain can effectively process the rest of the code. If everything still fits the screen. 

If it is again used at different places a good descriptive name is important.

That's the `Golang` naming philosophy.

File: src/test/Drips.t.sol [hunk 6]
```diff
@@ -917,4 +941,88 @@ contract DripsTest is DSTest, PseudoRandomUtils {
...
+        return Drips._defaultEnd(balance, list);
+    }
+
+    function testReceiverDefaultEndExampleA() public {
+        uint256 start1 = 50;
+        uint256 amtPerSec1 = 1;
+        uint256 start2 = 0;
+        uint256 amtPerSec2 = 1;
+
+        Hevm(HEVM_ADDRESS).warp(0);
```
[the remark this answers is not in the document]
Comment (xmxanuel) [reply src/test/Drips.t.sol:981]: Yes. I also thought of that.

File: src/Drips.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (xmxanuel) [reply src/Drips.sol:331]: The formular implemented has no logical meaning yet,  it is just the all balance spend equation expressed for x.

**All balance spend equation**
$$
a_1(x-s_1) + a_2(x-s_2) = b
$$

We can maybe spend some time to find some meaning why the implemented one  will exactly result in `defaultEnd`.

For now it is just the first equation expressed for `x`.

[the remark this answers is not in the document]
Comment (xmxanuel) [reply src/Drips.sol:331]: Ohh, or maybe I am wrong. Let me think.

[the remark this answers is not in the document]
Comment (CodeSandwich) [reply src/Drips.sol:331]: Ah, that's right, `defaultSum` is balance covering time between timestamp `0` and `start`, NOT `currTimestamp` and `start`!

File: src/Drips.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (CodeSandwich) [reply src/Drips.sol:304]: I'd say that `defaultSum` is better than `calcHelperSum` it's just as confusing but shorter.

When you read this code, this is where you see `defaultSum`:
```solidity
uint256 defaultSum = 0;
```
```solidity
defaultSum += uint256(receiver.amtPerSec) * start;
```
```solidity
 uint136 defaultEnd = uint136((balance + defaultSum) / defaultDivisor);
```
It's just 3 places and it's quite obvious what it holds: the total balance which would be dripped before each `start`. It's quite straight forward why it should be added to the `balance` when calculating when they all run out. `defaultSum` isn't an opaque result of a magical calculation, it's a simple and well defined part of an elegant algorithm. IMO it's much easier to understand it by following these 2 variables than by cracking the equation in your head and trying to match parts of it to variables.

File: src/Drips.sol [hunk 7]
```diff
@@ -307,78 +310,28 @@ library Drips {
             // Default drips end doesn't matter here, the end time is ignored when
             // the duration is zero and if it's non-zero the default end is not used anyway
             (uint32 start, uint32 end) = _dripsRangeInFuture(receiver, _currTimestamp(), 0);
+            if (start > highest) highest = start;
             if (receiver.duration == 0) {
```
[the remark this answers is not in the document]
Comment (xmxanuel) [reply src/Drips.sol]: Fixed.

File: src/test/Drips.t.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (xmxanuel) [reply src/test/Drips.t.sol]: Implemented it.

File: src/Drips.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (xmxanuel) [reply src/Drips.sol]: So you mean we can keep it as it is `defaultEnd`.

```

**Comment by mattosborn** [comment]
Love this! & it's exactly what I was wondering about in the meeting.

Another Q that came to mind:
- Is there any reason why `start` can't be used as the primary sort key for drips? Or does the storage update algorithm require `userId` be the primary sort key? & if so why is that?

**Comment by xmxanuel** [comment]
Hey @mattosborn 

We require it to be ordered by starting with `userId` for the storage update algorithm. 
Currently, the storage slot operations are the most expensive once. (adding start and end deltas).

We want to minimize the storage operations. 

How it works is that we have a `currReceiver` and a `newReceiver` list. If a sender wants to update his drips he proposes a `newReceiver` list. All drips which don't appear in the new list are removed, if start/end are the same we detect that and don't need to do anything otherwise it is an update. (or just new entries). 

You are right, that you can identify a lot of updates if both lists are just sorted in the same way.

However, the `userId` can't change so it makes sense to have it as the primary key.  If it changes it is basically another drip and never an update of an existing one. 

The problem with the `start` parameter is it can change.

For example, if I want to drip to you start: 10 and duration: 50. We would calculate the drip end at 60.

If I change my configuration now (before timestamp 10) to `start: 30, duration: 30`. 

The entry would be at a completely new position and in that case, the end time 60 didn't change. (So, we don't want to update the storage here).

**Comment by xmxanuel** [comment]
At least that's our reasoning here. 

However, I guess @CodeSandwich and I are obviously open for alternative ideas. Maybe we missed something over there is a even better way :)

File: src/Drips.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (xmxanuel) [reply src/Drips.sol]: I investigate why so many tests failed, when I changed that.

It is related to how we are performing the `assertBalance` check in the Drips test.

Image a case  we have a defaultDrip with 3 amtPerSec and we call the `assertBalance` check after the user hasn't enough balance to cover it anymore. 

Let's say the leftover balance is `1`.

The `assertBalance` basically results in a `setDrips` call where we pass the currReceiver as newReceiver. BalanceDelta=0. However, if it contains a defaultDrips with start=0 and duration=0. It will be interpreted as a new one. 

Then we end up the scenario where we don't have enough funds to cover the amtPerSec and the defaultEnd will in current `block.timestamp`. 

We could tolerate this user behavior. I think it would also happen if the user just wants to withdraw his balance but just passes his previous only defaultDrips. 

This would fix it.

```solidity
        require(defaultEnd > defaultsHighestStart || defaultEnd == block.timestamp, "Run out of funds before default drips start");
     
```
The alternative would be that we revert in such cases and adjust the tests.

wdyt @CodeSandwich

[the remark this answers is not in the document]
Comment (CodeSandwich) [reply src/Drips.sol]: I don't think that we can tolerate that scenario. How to describe this API behavior? "You must provide enough balance to cover at least 1 second of each unbound drips OR you must provide so little balance that you can't cover a single second of total amtPerSec of all unbound drips".

Another problem is that `defaultEnd != block.timestamp` if the uncovered unbound drip has a `start` in the future.

I think that we should adjust the tests. Maybe it'd be a good idea to add a balance calculating function to Drips lib and expose it in DripsHub? I think that it could be very useful for on-chain API users. Otherwise they'd need to duplicate drips logic on their side.
```solidity
function balanceAt(Storage storage s, uint256 userId, DripsReceiver[] memory receivers, uint32 timestamp) internal view returns (uint128 balance) { ... }
```

[the remark this answers is not in the document]
Comment (xmxanuel) [reply src/Drips.sol]: Change to `defaultEnd > defaultsHighestStart `.
Thanks for adding #163

**Review by CodeSandwich — COMMENTED** [review]
Looks really solid, I just have a few nitpicks :+1:

File: src/test/Drips.t.sol [hunk 8]
```diff
@@ -739,15 +780,25 @@ contract DripsTest is DSTest, PseudoRandomUtils {
...
+    }
+
     function testLimitsTheTotalReceiversCount() public {
         uint160 countMax = Drips.MAX_DRIPS_RECEIVERS;
         DripsReceiver[] memory receivers = new DripsReceiver[](countMax);
         for (uint160 i = 0; i < countMax; i++) {
             receivers[i] = recv(i, 1, 0, 0)[0];
         }
-        setDrips(sender, 0, 0, receivers);
+        setDrips(sender, 0, 1 ether, receivers);
```
Comment (CodeSandwich) [inline src/test/Drips.t.sol:793]:
```suggestion
        setDrips(sender, 0, 1 countMax, receivers);
```
Otherwise the the balance change below is invalid

**Review by CodeSandwich — APPROVED** [review]
Looks great :+1: I really like how you broke down the new algorithm, it's easy to understand and prove right!
